### DISCUSSION TITLE
Wait until element becomes clickable

### DISCUSSION THREAD

**Discussion by romek-codes** [seleniumbase/SeleniumBase#1332] [Q&A] [answered]
Is there a way to wait until element becomes clickable while using SeleniumBase?
I've only found a wait for checking presence or visibility but not "clickability".
I need it because while trying to click a Select element in Safari it is throwing me the "ElementNotInteractableException" yet it works in all other browsers.

**Comment by mdmintz** [comment, marked as answer]
@RomanJus That waiting is automatic with SeleniumBase. Is there more to the error message? (Eg. "Other element would receive the click: ..."). Are you running your SeleniumBase tests with ``--safari`` to get the native Safari integration? The SeleniumBase Safari integration includes some fixes that aren't part of the standard Selenium Safari APIs.

**Reply by romek-codes** [reply]
My stack call looks like this:
```
select_elements = sb.find_elements(EditProfilePage.WHO_ARE_YOU_SELECTS)
    
>       select_elements[0].click()

test_AllTests.py:793: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py:81: in click
    self._execute(Command.CLICK_ELEMENT)
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py:710: in _execute
    return self._parent.execute(command, params)
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py:425: in execute
    self.error_handler.check_response(response)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <selenium.webdriver.remote.errorhandler.ErrorHandler object at 0x11472ac50>
response = {'status': 400, 'value': '{"value":{"error":"element not interactable","message":"","stacktrace":""}}'}

    def check_response(self, response: Dict[str, Any]) -> None:
        """
        Checks that a JSON response from the WebDriver does not have an error.
    
        :Args:
         - response - The JSON response from the WebDriver server as a dictionary
           object.
    
        :Raises: If the response contains an error message.
        """
        if exception_class == UnexpectedAlertPresentException:
            alert_text = None
            if 'data' in value:
                alert_text = value['data'].get('text')
            elif 'alert' in value:
                alert_text = value['alert'].get('text')
            raise exception_class(message, screen, stacktrace, alert_text)  # type: ignore[call-arg]  # mypy is not smart enough here
>       raise exception_class(message, screen, stacktrace)
E       selenium.common.exceptions.ElementNotInteractableException: Message:

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py:247: ElementNotInteractableException
```
Indeed i'm running my tests with the --safari flag, this error happens for every select element that I have in my tests, and also only in Safari. I wasn't able to find a workaround yet.

**Reply by mdmintz** [reply]
@RomanJus Every time you call ``element.click()`` instead of ``self.click(selector)``, you're bypassing the built-in SeleniumBase functionality that's meant to help prevent these kinds of errors. The following example has a ``select`` element in it: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_demo_site.py (you can run that with ``--safari``).

**Reply by romek-codes** [reply]
Thanks, I will switch to using the SeleniumBase methods then.

**Reply by romek-codes** [reply]
Hi again, just wanted to say that i rewrote the function using the SeleniumBase methods but the Exception still happens on safari and safari only, i ran the demo test and it worked there.
I don't really know what else I can try and am running out of ideas for this problem so any direction you can point me towards would be  very appreciated.
I also wasn't able to find a SB method to check what the current selected option is, do you know what i could use for that?

Again, thank you for all your tips and help.

**Reply by mdmintz** [reply]
Hi @RomanJus,
To get the current value of the ``select``, you can use:
``self.get_property("select", "value")``
(If there's more than one ``select`` on that page, you'll need to use a more-specific selector.)
Here's an example of changing the ``select``:
``self.select_option_by_text("#mySelect", "Set to 75%")``
(https://github.com/seleniumbase/SeleniumBase/blob/master/examples/test_demo_site.py)

**Reply by xlapsed** [reply]
Is there any shortcut to get the text of the current select option?  I'm currently using self.get_property to retrieve the value, then look up the text based on the value, but I was hoping there was a more direct way to just get the text.

self.get_property("selector", "text") doesn't seem to work.

**Reply by mdmintz** [reply]
@ken-smith-dev There's `value` for `get_property`, but if you want the text of something, use `self.get_text(selector)` or `self.get_text_content(selector)`. There's also a `self.get_property_value(selector, property)` method, which is different from `self.get_property(selector, property`. (See [seleniumbase/fixtures/base_case.py](https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/fixtures/base_case.py) to know the difference.)

**Comment by wackydude** [comment]
Issue I am having is when I type into email input field on a shipping page, once the field contains a valid email, the page grays out and an animation runs for about 3 seconds while a call is made to get address specific shipping methods. When the call returns, the appropriate shipping methods are added to the page with default method auto-selected, and the 'continue to payment' button updates from grayed out to enabled as expected, however when I try sb.click(continue_to_payment_btn, scroll=False) it is not being clicked. If I add sb.sleep(3) at the end of the method that types the email_address into the field, then the sb.click(continue_to_payment_btn, scroll=False) works. So I believe you that sbase should automatically wait for things to become clickable, but in this case there seems to be an issue related to either (or both) of loading animations or elements just added to the page and becoming enabled. I even have tried sb.wait_for_element(continue_to_payment_btn, timeout=10) before trying to click and that doesn't fix the problem, but a hardcoded 3 second wait after entering the email while the animation displays does work. Hope it helps. Thanks for a great tool!
